### PR TITLE
Configuring for pure-python

### PR DESCRIPTION
**Pull request by mgedmin**
This does:
- bin/python config-package.py ~/src/zopefoundation/zope.testrunner --type pure-python --with-appveyor --with-pypy --with-docs
- some edits to `.meta.toml` for things that cannot be specified on the command line, like fail_under = 85 and additional flake8/tox/appveyor options
- rerunning bin/python config-package.py again
- adding 3.9 classifier to setup.py to make check-python-versions-happy
- fixes subprocess support on Windows, which I suspect was always broken

It drops some coverage-related configuration, but passes the 85% threshold despite of that:
- we're missing some exclusions from the old .coveragerc
- we're now enabling branch coverage
- we're no longer doing COVERAGE_PROCESS_START (and `coverage.process_startup()` invocations from the setup.py-test-generated wrapper script)

TL;DR: we now use GitHub Actions instead of Travis CI, and we no longer use `setup.py test`, and running layers in a subprocess now works under Windows.

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -1,20 +0,0 @@
-[run]
-source = zope.testrunner
-parallel = true
-omit =
-    # These files are definitely imported and executed,
-    # but for some reason don't show up as such.
-    # See testrunner-knit
-    src/zope/testrunner/tests/testrunner-ex-pp-products/more/sampletests.py
-    src/zope/testrunner/tests/testrunner-ex-pp-products/sampletests.py
```
Comment (mgedmin) [inline .coveragerc:9]: I suppose these need to be moved into `meta.toml` and regenerated into wherever that template puts coverage configs

File: .coveragerc [hunk 2]
```diff
@@ -1,20 +0,0 @@
...
-[report]
-precision = 2
-exclude_lines =
-    pragma: no cover
-    if __name__ == '__main__':
-    raise NotImplementedError
-    self.fail
-    raise AssertionError
-    if is_jython:
-    if sys.platform == 'win32':
```
Comment (mgedmin) [inline .coveragerc:20]: This also I'm sure includes things not mentioned in the standard template.

File: .meta.toml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+[python]
+with-appveyor = true
+with-pypy = true
+with-legacy-python = true
+with-docs = false
+with-sphinx-doctests = false
+
+[coverage]
+fail-under = 0
```
Comment (mgedmin) [inline .meta.toml:15]: This needs fixing

File: .travis.yml [hunk 4]
```diff
@@ -1,25 +0,0 @@
...
-env:
-  - SUBUNIT=true
-  - SUBUNIT=false
-install:
-    - pip install -U pip setuptools
-    - pip install -U coverage coveralls
-    - pip install -U -e .[test]
-    - test "$SUBUNIT" != true || pip install -U -e .[subunit]
-script:
-    - COVERAGE_PROCESS_START=`pwd`/.coveragerc coverage run setup.py -q test -q
```
Comment (mgedmin) [inline .travis.yml:19]: Oof, I suppose this will need moving into tox.ini

File: appveyor.yml [hunk 5]
```diff
@@ -1,7 +1,7 @@
+# Generated from:
+# https://github.com/zopefoundation/meta/tree/master/config/pure-python
 environment:
 
-  SUBUNIT: 0
-
```
Comment (mgedmin) [inline appveyor.yml:4]: This will need to be moved into the template

File: appveyor.yml [hunk 6]
```diff
@@ -13,29 +13,30 @@ environment:
...
-    - { python: 27, SUBUNIT: 1 }
-    - { python: 27-x64, SUBUNIT: 1 }
-    - { python: 35, SUBUNIT: 1 }
-    - { python: 35-x64, SUBUNIT: 1 }
-    - { python: 36, SUBUNIT: 1 }
-    - { python: 36-x64, SUBUNIT: 1 }
-    - { python: 37, SUBUNIT: 1 }
-    - { python: 37-x64, SUBUNIT: 1 }
-    - { python: 38, SUBUNIT: 1 }
-    - { python: 38-x64, SUBUNIT: 1 }
```
Comment (mgedmin) [inline appveyor.yml:25]: These will need to be resurrected somehow

File: appveyor.yml [hunk 7]
```diff
@@ -13,29 +13,30 @@ environment:
...
+      if (-not (Test-Path $env:PYTHON)) {
+        curl -o install_python.ps1 https://raw.githubusercontent.com/matthew-brett/multibuild/11a389d78892cf90addac8f69433d5e22bfa422a/install_python.ps1
+        .\install_python.ps1
+      }
+  - ps: if (-not (Test-Path $env:PYTHON)) { throw "No $env:PYTHON" }
   - echo "C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin\SetEnv.cmd" /x64 > "C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\bin\amd64\vcvars64.bat"
-  - python -m pip install -U setuptools pip
-  # NB: pip install -e .[test] fails for obscure namespace package reasons
-  - pip install -U .[test]
-  - if %SUBUNIT%==1 pip install -U .[subunit]
```
Comment (mgedmin) [inline appveyor.yml:33]: Oof this too

File: appveyor.yml [hunk 8]
```diff
@@ -13,29 +13,30 @@ environment:
...
+  - pip install -U setuptools wheel
+  - pip install -U -e .[test]
 
 build: false
+matrix:
+  fast_finish: true
 
 test_script:
-  - python setup.py -q test -q
+  - zope-testrunner --test-path=src
```
Comment (mgedmin) [inline appveyor.yml:39]: Hmm.  Can we rely on zope.testrunner to run its own tests without relying on whatever hacks `setup.py test` does?

File: tox.ini [hunk 9]
```diff
@@ -1,35 +1,69 @@
...
+    lint
+    py27
+    py35
+    py36
+    py37
+    py38
+    py39
+    pypy
+    pypy3
+    coverage
```
Comment (mgedmin) [inline tox.ini:15]: Yeah this drops docs and all the subunit variants, not good.

File: tox.ini [hunk 10]
```diff
@@ -1,35 +1,69 @@
...
 
 [testenv]
+usedevelop = true
+deps =
+    zope.testrunner
+commands =
+    zope-testrunner --test-path=src {posargs:-vc}
 extras =
     test
-    subunit: subunit
```
Comment (mgedmin) [inline tox.ini:8]: We need to keep this bit.

File: tox.ini [hunk 11]
```diff
@@ -1,35 +1,69 @@
...
+    check-python-versions
 
 [testenv:coverage]
-usedevelop = true
-basepython =
-    python3
-commands =
-    coverage run setup.py -q test -q
-    coverage combine
-    coverage report --fail-under=85
```
Comment (mgedmin) [inline tox.ini:19]: Number!

File: tox.ini [hunk 12]
```diff
@@ -1,35 +1,69 @@
...
-setenv =
-    COVERAGE_PROCESS_START = {toxinidir}/.coveragerc
```
Comment (mgedmin) [inline tox.ini:21]: We probably need to keep this

File: tox.ini [hunk 13]
```diff
@@ -1,35 +1,69 @@
...
-extras =
-    test
-    subunit
```
Comment (mgedmin) [inline tox.ini:24]: and this

File: tox.ini [hunk 14]
```diff
@@ -1,35 +1,69 @@
...
 deps =
     coverage
-
-[testenv:docs]
-basepython =
-    python3.6
+    coverage-python-version
+    zope.testrunner
 commands =
-    sphinx-build -b html -d docs/_build/doctrees docs docs/_build/html
```
Comment (mgedmin) [inline tox.ini:32]: And I totally forgot to enable sphinx docs when running the meta script.

**Review by mgedmin — COMMENTED** [review]
So this is a larger job than I initially realized.

I'm mostly worried about the chicken-and-egg situation of testing zope.testrunner using zope.testrunner.  Can we drop setup.py test or does it do something essential to enable bootstrapping?

**Comment by jamadden** [comment]
Is zope.testrunner special enough that it's not worth it to "standardize" with this tool? Setting up a GHA .yml file that replicates what .travis.yml did for pure-python projects is relatively straightforward-ish  (not that I'm volunteering right now 😄), though of course the more complicated .travis.yml is, the worse the GHA will always be.

**Comment by cjwatson** [comment]
It's not really a chicken-and-egg question given that `setup.py test` uses `zope.testrunner` internally anyway, just a question of whether its extra hacks matter.

I think we should give removing it a try.  I killed off just `setup.py test` locally without doing any of the rest of the stuff in this PR (except for updating `tox.ini`) and `tox` seems happy locally except for maybe a problem with insufficient coverage; if there are still any CI problems with that then we should probably investigate them harder.

**Comment by icemac** [comment]
When #117 is merged GHA should at least run the tests.

File: .meta.toml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+# Generated from:
+# https://github.com/zopefoundation/meta/tree/master/config/pure-python
+[meta]
+template = "pure-python"
```
Comment (icemac) [inline .meta.toml:4]: Maybe the template `buildout-recipe` is better suited here as it provides infrastructure for coverage when using multiple processes.

Comment (mgedmin) [reply .meta.toml:4]: Nah, zope.testrunner is not a buildout recipe.  I'd rather add a flag to the pure-python recipe for supporting multiprocess coverage, I expect it'll come in handy for other packages (e.g. zdaemon).

**Comment by cjwatson** [comment]
> When #117 is merged GHA should at least run the tests.

This is done now, so worth rebasing I think.

**Comment by mgedmin** [comment]
PR updated, still very not ready.

File: appveyor.yml (hunk 8 above)
Comment (mgedmin) [reply appveyor.yml:39]: The consensus seem to be "probably"?

File: tox.ini (hunk 9 above)
Comment (mgedmin) [reply tox.ini:16]: Semi-fixed, but I'm now enumerating all the supported Python versions in .meta.toml, in `additional-envlist`.

File: tox.ini (hunk 10 above)
Comment (mgedmin) [reply tox.ini:8]: Fix relies on https://github.com/zopefoundation/meta/pull/71.

File: .github/workflows/tests.yml [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        - ["2.7",   "py27"]
+        - ["3.5",   "py35"]
+        - ["3.6",   "py36"]
+        - ["3.7",   "py37"]
+        - ["3.8",   "py38"]
+        - ["3.9",   "py39"]
+        - ["pypy2", "pypy"]
+        - ["pypy3", "pypy3"]
+        - ["3.8",   "docs"]
+        - ["3.8",   "coverage"]
```
Comment (mgedmin) [inline .github/workflows/tests.yml:28]: Oops, need to add all the subunit variants here.

File: setup.cfg [hunk 17]
```diff
@@ -1,7 +1,13 @@
+# Generated from:
+# https://github.com/zopefoundation/meta/tree/master/config/pure-python
 [bdist_wheel]
 universal = 1
 
 [flake8]
+doctests = 1
```
Comment (mgedmin) [inline setup.cfg:7]: Oh, I've only just now realized the additional flake8 errors I had to additionally fix came from `doctests = 1` here!

File: tox.ini [hunk 18]
```diff
@@ -1,43 +1,81 @@
...
-    subunit: subunit
-commands =
-    python setup.py -q test -q
-
-[testenv:coverage]
 usedevelop = true
-basepython =
-    python3
+deps =
+    zope.testrunner
```
Comment (mgedmin) [inline tox.ini:22]: I wonder if we should drop this?  Kind of pointless to install zope.testrunner from pypi before reinstalling it from the local source tree.

(This would need another template tweak.)

**Review by mgedmin — COMMENTED** [review]
This no longer feels a hopeless task!

File: tox.ini (hunk 13 above)
Comment (mgedmin) [reply tox.ini:24]: this could also explain the coverage percentage drop!

Comment (mgedmin) [reply tox.ini:24]: huh, re-enabling the `subunit` extra for tox -e coverage brought our percentage back to 85%!

**Comment by mgedmin** [comment]
I would prefer to get https://github.com/zopefoundation/meta/pull/71 merged before I create a new PR for tweaking appveyor.yml.

There currently is an escape hatch, but I don't want to wrap the entire appveyor.yml inside a TOML list, I only need a few tweaks (global envvars, and additional matrix rows)

**Comment by mgedmin** [comment]
Ok, current status: appveyor is broken as heck, and I'm not sure why.

File: .meta.toml (hunk 15 above)
Comment (icemac) [reply .meta.toml:4]: At least in the case of buildout recipes the coverage config has to be stored in a `.coveragerc` – that's the main difference to the `pure-python` template. Having both `.coveragerc` and coverage config in `tox.ini` in a template makes it even more complex.
Maybe the name of the `buildout-recipe` template is only too specific and it should be called `coverage-over-subprocesses` (?).

File: tox.ini (hunk 18 above)
Comment (icemac) [reply tox.ini:22]: I think it would be worth the additional template tweak.

**Comment by mgedmin** [comment]
Ok, so the Windows breakage: zope-testrunner cannot run tests in a subprocess because it tries to execute `/python27/python.exe /python27/scripts/zope-testrunner`, and python.exe says `/python27/scripts/zope-testrunner` does not exist.

Before, we had this interesting bit:
https://github.com/zopefoundation/zope.testrunner/blob/9b7cc179ee857e434f85d7e1cdc873ef4ded1bd0/src/zope/testrunner/__init__.py#L63-L64

i.e. I think we've been running `/python27/python.exe -c "from zope.testrunner import run; run()"`.

I wonder if we could always do [sys.executable, `-mzope.testrunner`] instead?

**Comment by mgedmin** [comment]
I don't have a Windows machine to test, or, rather, I\'m not going to dual-boot into Windows on this laptop.  IIRC the 'scripts/' folder gets three wrapper scripts per console_script entry, one with .py, one with .exe, and the third one is some weird  -script.exe?  Or -wrapper.exe?  I honestly don't remember.

I like the idea of doing `python -m zope.testrunner` instead of trying to figure out what's the right wrapper script filename to pass to python.exe.  I'm only going to do that if the current logic looking for the wrapper script fails to find one, so this should not regress on Linux or MacOS.  And AFAIU things on Windows were already broken wrt subprocesses, except for that one case where you were using `python setup.py test` as a wrapper for zope.testrunner.

So, I think this is finally in a mergeable state.  I don't have the energy to see if I can improve coverage numbers by tweaking coverage configs.  I might find the energy to tweak the meta config template to drop the zope.testrunner self-dependency in tox.ini, especially if somebody gently prods me to do it.

**Comment by mgedmin** [comment]
It seems that I ran out of steam for the nice-to-have improvements (deps in tox.ini, investigating coverage numbers).

I propose merging this as-is, unless anyone objects?.

**Comment by icemac** [comment]
This PR creates a much better situation than the one we currently have on `master`. Maybe you could create an issue so we do not forget the nice-to-have improvements and can tackle them later on.

File: tox.ini (hunk 18 above)
Comment (mgedmin) [reply tox.ini:22]: https://github.com/zopefoundation/meta/pull/85

**Comment by mgedmin** [comment]
> Maybe you could create an issue so we do not forget the nice-to-have improvements and can tackle them later on.

https://github.com/zopefoundation/zope.testrunner/pull/118
